pgml-extension/src/orm/snapshot.rs

```diff
@@ -811,19 +829,217 @@ impl Snapshot {
+                        if i < num_train_rows {
+                            &mut text2_train
+                        } else {
+                            &mut text2_test
+                        }
+                    } else if column.name == class_column_value {
+                        if i < num_train_rows {
+                            &mut class_train
+                        } else {
+                            &mut class_test
+                        }
+                    } else {
+                        continue;
+                    };
+
+                    match column.pg_type.as_str() {
+                        "bpchar" | "text" | "varchar" => match row[column.position].value::<String>().unwrap() {
+                            Some(text) => vector.push(text),
+                            None => error!("NULL training text is not handled"),
+                        },
+                        _ => error!("only text type columns are supported"),
+                    }
+                }
+            });
+
+            let num_distinct_labels = class_train.iter().cloned().collect::<HashSet<_>>().len();
+            data = Some(TextPairClassificationDataset {
+                text1_train,
+                text2_train,
+                class_train,
+                text1_test,
+                text2_test,
+                class_test,
                 num_features,
                 num_labels,
                 num_rows,
                 num_test_rows,
                 num_train_rows,
                 // TODO rename and audit this
-                num_distinct_labels: self.num_classes(),
+                num_distinct_labels,
+            });
+
+            Ok::<std::option::Option<()>, i64>(Some(())) // this return type is nonsense
+        })
+        .unwrap();
+
+        let data = data.unwrap();
+
+        info!("{}", data);
+
+        data
+    }
+
+    pub fn conversation_dataset(&mut self, dataset_args: default!(JsonB, "'{}'")) -> ConversationDataset {
+        let mut data = None;
+
+        Spi::connect(|client| {
+            let result = client.select(&self.select_sql(), None, None).unwrap();
+            let num_rows = result.len();
+            let (num_train_rows, num_test_rows) = self.train_test_split(num_rows);
+            let num_features = 2;
+
+            let mut system_train: Vec<String> = Vec::with_capacity(num_train_rows);
+            let mut user_train: Vec<String> = Vec::with_capacity(num_train_rows);
+            let mut assistant_train: Vec<String> = Vec::with_capacity(num_train_rows);
+            let mut system_test: Vec<String> = Vec::with_capacity(num_test_rows);
+            let mut user_test: Vec<String> = Vec::with_capacity(num_test_rows);
+            let mut assistant_test: Vec<String> = Vec::with_capacity(num_test_rows);
+
+            let system_column_value = dataset_args
+                .0
+                .get("system_column")
```

Comment:
How standard are these names these days?

Reply:
For conversation task, system, user and assistant have become standard keys.